### PR TITLE
spirv-fuzz: Support atomic operations opcode

### LINKED ISSUE #4345: spirv-fuzz: Allow different signedness in scope and semantics arguments for atomic operations
Many atomic operations take a Scope argument, and a Memory Semantics argument.

These must be the ids of 32-bit integers, but their signedness does not matter.

`TransformationReplaceIdWithSynonym` has a method, `IsAgnosticToSignednessOfOperand`, which should return true to indicate that the signedness of a particular operand is unimportant. The instructions handled by this method are not complete, and in particular all of the atomic operations are missing.

This issue involves enhancing `TransformationReplaceIdWithSynonym::IsAgnosticToSignednessOfOperand` so that it correctly returns `true` for all of the scope and memory semantics operands of relevant instructions, from these parts of the SPIR-V specification:

https://www.khronos.org/registry/spir-v/specs/unified1/SPIRV.html#Memory_Semantics_-id-

https://www.khronos.org/registry/spir-v/specs/unified1/SPIRV.html#Scope_-id-

Furthermore, test cases should be added to cover a reasonable number of these cases - all the cases that are easy to express as Vulkan compute shaders should be covered.

Related issue: #3582.

### PR DESCRIPTION
**Pull request by Mostafa-ashraf19**
Fixes #4345. 

This PR is responsible for supporting atomic operations opcode and checks it's in operands signedness neutrality.

### Example:

Original shader 

https://github.com/KhronosGroup/SPIRV-Tools/blob/8cc8b6562be9de063a22ddf7f75d11c641779e36/test/fuzz/transformation_replace_id_with_synonym_test.cpp#L1911-L1948

After applying transformation

https://github.com/KhronosGroup/SPIRV-Tools/blob/8cc8b6562be9de063a22ddf7f75d11c641779e36/test/fuzz/transformation_replace_id_with_synonym_test.cpp#L1992-L2029

That's the Important changes from original shader ` %21 = OpAtomicLoad %6 %14 %15 %20`, and this transformed  shader ` %21 = OpAtomicLoad %6 %14 %100 %101`.

### CODE REVIEW

File: source/fuzz/transformation_replace_id_with_synonym.cpp [hunk 1]
```diff
@@ -139,6 +139,26 @@ bool TransformationReplaceIdWithSynonym::IsAgnosticToSignednessOfOperand(
...
+    case SpvOpAtomicUMin:
+    case SpvOpAtomicSMax:
+    case SpvOpAtomicUMax:
+    case SpvOpAtomicAnd:
+    case SpvOpAtomicOr:
+    case SpvOpAtomicXor:
+    // These instructions required another capability.
+    case SpvOpAtomicFlagTestAndSet:
+    case SpvOpAtomicFlagClear:
+    case SpvOpAtomicFAddEXT:
```
Comment (afd) [inline source/fuzz/transformation_replace_id_with_synonym.cpp:161]: This is the right function to be editing, but what you have done is not quite enough.

For the existing instructions above, such as `OpSGreaterThanEqual`, all of their arguments have integer type and the signedness is not important, so we can ignore the `use_in_operand_index` parameter.

However, that's not true for atomic operations.

For example, with `OpAtomicLoad` there are three operands: pointer, memory scope and memory semantics. The instruction is agnostic to the signedness of the memory scope and memory semantics operands. So we should return true for this instruction only when `use_in_operand_index` is >= 1.

With `OpAtomicCompareExchange` there are several operands: pointer, memory scope, memory semantics for equality, memory semantics for inequality, value, and comparator. If you read the description of that instruction carefully you'll see that the value and comparator operands have to have types that match the result type of the instruction, and the pointee type of the pointer, respectively, thus the instruction is *not* agnostic to the signedness of these operands. So for this instruction we should return true only if `use_in_operand_index` is 1, 2 or 3.

Can you please go through the instructions and think carefully about for which operands the signedness doesn't matter?

**Review by afd — CHANGES_REQUESTED** [review]
Thanks for making a start. However, more care is required to identify exactly which operands of these new instructions are agnostic to signedness - it's not true for all operands.

Also, you should enable the disabled test associated with this issue, and enhance that test to feature examples for each of the new instructions you've considered here. If you can try to cover them all in the test that would be great - or at least the ones that use Shader capability.

File: source/fuzz/transformation_replace_id_with_synonym.cpp (hunk 1 above)
Comment (Mostafa-ashraf19) [reply source/fuzz/transformation_replace_id_with_synonym.cpp:161]: Okay, I will do that.

File: test/fuzz/transformation_replace_id_with_synonym_test.cpp [hunk 2]
```diff
@@ -1888,7 +1888,7 @@ TEST(TransformationReplaceIdWithSynonymTest,
 //  test so that it covers more atomic operations, and enable the test once the
 //  issue is fixed.
 TEST(TransformationReplaceIdWithSynonymTest,
```
Comment (afd) [inline test/fuzz/transformation_replace_id_with_synonym_test.cpp:1890]: Because this PR addresses the above TODO comment, the TODO comment should be removed as part of the PR.

File: test/fuzz/transformation_replace_id_with_synonym_test.cpp [hunk 3]
```diff
@@ -1888,7 +1888,7 @@ TEST(TransformationReplaceIdWithSynonymTest,
...
-     DISABLED_SignOfAtomicScopeAndMemorySemanticsDoesNotMatter) {
+     SignOfAtomicScopeAndMemorySemanticsDoesNotMatter) {
   // TODO(https://github.com/KhronosGroup/SPIRV-Tools/issues/4345): both the
   //  GLSL comment and the corresponding SPIR-V should be updated to cover a
   //  larger number of atomic operations.
```
Comment (afd) [inline test/fuzz/transformation_replace_id_with_synonym_test.cpp:1894]: I think we need to keep this test disabled at the moment.

The reason is that #4349 is going to make it illegal to replace a memory semantics or memory scope operand, so replacing them with constants of a different sign won't be allowed anymore by spirv-fuzz (even though it is legal).

We can consider returning to this particular test in the future.

File: test/fuzz/transformation_replace_id_with_synonym_test.cpp [hunk 4]
```diff
@@ -2034,7 +2034,7 @@ TEST(TransformationReplaceIdWithSynonymTest,
 // TODO(https://github.com/KhronosGroup/SPIRV-Tools/issues/4345): Improve this
 //  test so that it covers more atomic operations, and enable the test once the
 //  issue is fixed.
```
Comment (afd) [inline test/fuzz/transformation_replace_id_with_synonym_test.cpp:2036]: Please remove this comment as part of the PR, since the purpose of the PR is to enable the functionality that this tests.

File: test/fuzz/transformation_replace_id_with_synonym_test.cpp [hunk 5]
```diff
@@ -2046,8 +2046,20 @@ TEST(TransformationReplaceIdWithSynonymTest, DISABLED_TypesAreCompatible) {
...
+         %13 = OpTypePointer StorageBuffer %6
+         %15 = OpConstant %6 2
+         %16 = OpConstant %6 7
+         %20 = OpConstant %9 64
           %4 = OpFunction %2 None %3
           %5 = OpLabel
+         %14 = OpAccessChain %13 %11 %12
+         %21 = OpAtomicLoad %6 %14 %15 %20
+         %22 = OpAtomicExchange %6 %14 %15 %20 %16
+               OpAtomicStore %14 %15 %20 %16
```
Comment (afd) [inline test/fuzz/transformation_replace_id_with_synonym_test.cpp:2062]: Please improve this, and the rest of the test, to cover all the atomic operations that are allowed with the Shader capability.

**Review by afd — CHANGES_REQUESTED** [review]
Thanks for the progress on this - some comments inline.

File: test/fuzz/transformation_replace_id_with_synonym_test.cpp (hunk 2 above)
Comment (Mostafa-ashraf19) [reply test/fuzz/transformation_replace_id_with_synonym_test.cpp:1890]: Done.

File: test/fuzz/transformation_replace_id_with_synonym_test.cpp (hunk 4 above)
Comment (Mostafa-ashraf19) [reply test/fuzz/transformation_replace_id_with_synonym_test.cpp:2036]: Done.

File: test/fuzz/transformation_replace_id_with_synonym_test.cpp (hunk 3 above)
Comment (Mostafa-ashraf19) [reply test/fuzz/transformation_replace_id_with_synonym_test.cpp:1894]: All right!

File: test/fuzz/transformation_replace_id_with_synonym_test.cpp (hunk 5 above)
Comment (Mostafa-ashraf19) [reply test/fuzz/transformation_replace_id_with_synonym_test.cpp]: Done.

File: source/fuzz/transformation_replace_id_with_synonym.cpp [hunk 6]
```diff
@@ -140,6 +140,51 @@ bool TransformationReplaceIdWithSynonym::IsAgnosticToSignednessOfOperand(
...
+    case SpvOpAtomicXor:
+    // TBD(To Be Discussed),
+    // https://github.com/KhronosGroup/SPIRV-Registry/blob/master/extensions/EXT/SPV_EXT_shader_atomic_float_add.asciidoc#modifications-to-the-spir-v-specification-version-15
+    case SpvOpAtomicFAddEXT:  // Capability (AtomicFloat32AddEXT,
+                              // AtomicFloat64AddEXT)
+      assert(use_in_operand_index != 0 && "Forbidden.");
+
+      if (opcode == SpvOpAtomicFAddEXT) {
+        // Would you like to check Capability here?
+      }
```
Comment (afd) [inline source/fuzz/transformation_replace_id_with_synonym.cpp:163]: You can include that instruction in the list of case labels, and there's no need to check for capability.

The reason is that whether signedness of operands matters is independent of the capability.

That's different from the situation with operands being OpConstant in your previous PR - there, the capability determined the rules. Here the capability only determines whether this instruction exists. If it doesn't exist we won't see it in this function, so that's OK :)

In short: you can just include it in the list of case labels, and then have your `return` statement handle all of them.

File: source/fuzz/transformation_replace_id_with_synonym.cpp [hunk 7]
```diff
@@ -140,6 +140,51 @@ bool TransformationReplaceIdWithSynonym::IsAgnosticToSignednessOfOperand(
...
+      return use_in_operand_index == 1 || use_in_operand_index == 2;
+
+    case SpvOpAtomicCompareExchange:
+    // Deprecated, missing after version 1.3.
+    // Would you like to remove it?
```
Comment (afd) [inline source/fuzz/transformation_replace_id_with_synonym.cpp:168]: Remove this.

File: source/fuzz/transformation_replace_id_with_synonym.cpp [hunk 8]
```diff
@@ -140,6 +140,51 @@ bool TransformationReplaceIdWithSynonym::IsAgnosticToSignednessOfOperand(
...
+    case SpvOpAtomicSMax:
+    case SpvOpAtomicUMax:
+    case SpvOpAtomicAnd:
+    case SpvOpAtomicOr:
+    case SpvOpAtomicXor:
+    // TBD(To Be Discussed),
+    // https://github.com/KhronosGroup/SPIRV-Registry/blob/master/extensions/EXT/SPV_EXT_shader_atomic_float_add.asciidoc#modifications-to-the-spir-v-specification-version-15
+    case SpvOpAtomicFAddEXT:  // Capability (AtomicFloat32AddEXT,
+                              // AtomicFloat64AddEXT)
+      assert(use_in_operand_index != 0 && "Forbidden.");
```
Comment (afd) [inline source/fuzz/transformation_replace_id_with_synonym.cpp:159]: Change "Forbidden" to something like: "Signedness check should not occur on a pointer operand."

File: source/fuzz/transformation_replace_id_with_synonym.cpp [hunk 9]
```diff
@@ -140,6 +140,51 @@ bool TransformationReplaceIdWithSynonym::IsAgnosticToSignednessOfOperand(
...
+      if (opcode == SpvOpAtomicFAddEXT) {
+        // Would you like to check Capability here?
+      }
+      return use_in_operand_index == 1 || use_in_operand_index == 2;
+
+    case SpvOpAtomicCompareExchange:
+    // Deprecated, missing after version 1.3.
+    // Would you like to remove it?
+    case SpvOpAtomicCompareExchangeWeak:  // Capability (Kernel)
+      assert(use_in_operand_index != 0 && "Forbidden.");
```
Comment (afd) [inline source/fuzz/transformation_replace_id_with_synonym.cpp:170]: Similar.

File: source/fuzz/transformation_replace_id_with_synonym.cpp [hunk 10]
```diff
@@ -140,6 +140,51 @@ bool TransformationReplaceIdWithSynonym::IsAgnosticToSignednessOfOperand(
...
+      // This opcode is Deprecated!
+      if (opcode == SpvOpAtomicCompareExchangeWeak) {
+        // Would you like to check Capability here?
+      }
```
Comment (afd) [inline source/fuzz/transformation_replace_id_with_synonym.cpp:174]: Same as above, there's no need for this check. If the module contains an OpAtomicCompareExchangeWeak instruction then we should handle signedness of its operands appropriately.

File: source/fuzz/transformation_replace_id_with_synonym.cpp [hunk 11]
```diff
@@ -140,6 +140,51 @@ bool TransformationReplaceIdWithSynonym::IsAgnosticToSignednessOfOperand(
...
+      return use_in_operand_index >= 1 && use_in_operand_index <= 3;
+    case SpvOpAtomicLoad:
+    case SpvOpAtomicIIncrement:
+    case SpvOpAtomicIDecrement:
+    case SpvOpAtomicFlagTestAndSet:  // Capability (Kernel)
+    case SpvOpAtomicFlagClear:       // Capability (Kernel)
+      assert(use_in_operand_index != 0 && "Forbidden.");
```
Comment (afd) [inline source/fuzz/transformation_replace_id_with_synonym.cpp:181]: Similar change to the string.

File: source/fuzz/transformation_replace_id_with_synonym.cpp [hunk 12]
```diff
@@ -140,6 +140,51 @@ bool TransformationReplaceIdWithSynonym::IsAgnosticToSignednessOfOperand(
...
+      if (opcode == SpvOpAtomicFlagTestAndSet ||
+          opcode == SpvOpAtomicFlagClear) {
+        // Would you like to check Capability here?
+      }
```
Comment (afd) [inline source/fuzz/transformation_replace_id_with_synonym.cpp:185]: No need for these checks.

File: test/fuzz/transformation_replace_id_with_synonym_test.cpp (hunk 2 above)
Comment (afd) [reply test/fuzz/transformation_replace_id_with_synonym_test.cpp:1887]: Actually, I think now that 4349 is merged this test will no longer pass: 4349 will stop the synonym replacements being possible due to the rules on constants for the Shader capability.

Please confirm this by updating master and rebasing, and then seeing whether this fails. If it does fail, please put back the comment about the test being disabled and make the test disabled again.

If it passes then let's discuss (because I don't think it should).

**Review by afd — CHANGES_REQUESTED** [review]
Some clarifications.

File: source/fuzz/transformation_replace_id_with_synonym.cpp (hunk 8 above)
Comment (Mostafa-ashraf19) [reply source/fuzz/transformation_replace_id_with_synonym.cpp:159]: Done.

File: source/fuzz/transformation_replace_id_with_synonym.cpp (hunk 7 above)
Comment (Mostafa-ashraf19) [reply source/fuzz/transformation_replace_id_with_synonym.cpp:168]: Done.

File: source/fuzz/transformation_replace_id_with_synonym.cpp (hunk 9 above)
Comment (Mostafa-ashraf19) [reply source/fuzz/transformation_replace_id_with_synonym.cpp:170]: Done.

File: source/fuzz/transformation_replace_id_with_synonym.cpp (hunk 10 above)
Comment (Mostafa-ashraf19) [reply source/fuzz/transformation_replace_id_with_synonym.cpp:174]: OK, Done.

File: source/fuzz/transformation_replace_id_with_synonym.cpp (hunk 12 above)
Comment (Mostafa-ashraf19) [reply source/fuzz/transformation_replace_id_with_synonym.cpp:185]: OK, removed.

File: source/fuzz/transformation_replace_id_with_synonym.cpp (hunk 11 above)
Comment (Mostafa-ashraf19) [reply source/fuzz/transformation_replace_id_with_synonym.cpp:181]: Changed.

File: source/fuzz/transformation_replace_id_with_synonym.cpp (hunk 6 above)
Comment (Mostafa-ashraf19) [reply source/fuzz/transformation_replace_id_with_synonym.cpp:163]: Thanks for the clarification, done.

File: test/fuzz/transformation_replace_id_with_synonym_test.cpp (hunk 2 above)
Comment (Mostafa-ashraf19) [reply test/fuzz/transformation_replace_id_with_synonym_test.cpp:1887]: @afd Test `SignOfAtomicScopeAndMemorySemanticsDoesNotMatter` fails.

File: test/fuzz/transformation_replace_id_with_synonym_test.cpp [hunk 13]
```diff
@@ -2026,11 +2026,10 @@ TEST(TransformationReplaceIdWithSynonymTest,
                    .IsApplicable(context.get(), transformation_context));
 }
 
-// TODO(https://github.com/KhronosGroup/SPIRV-Tools/issues/4345): Improve this
-//  test so that it covers more atomic operations, and enable the test once the
-//  issue is fixed.
```
Comment (afd) [inline test/fuzz/transformation_replace_id_with_synonym_test.cpp:2031]: You should reinstate this comment, since we weren't able to enable the test after all.

File: test/fuzz/transformation_replace_id_with_synonym_test.cpp [hunk 14]
```diff
@@ -2026,11 +2026,10 @@ TEST(TransformationReplaceIdWithSynonymTest,
...
 TEST(TransformationReplaceIdWithSynonymTest,
      DISABLED_SignOfAtomicScopeAndMemorySemanticsDoesNotMatter) {
+  //  This test is failed at (IsApplicable) line, because of updates from PR
+  //  #4349.
```
Comment (afd) [inline test/fuzz/transformation_replace_id_with_synonym_test.cpp:2032]: This kind of comment doesn't belong in the source code, and we already have TODO comments explaining that the test isn't expected to pass at present, so please remove.

**Review by afd — CHANGES_REQUESTED** [review]
Looks good - just a couple of comment-related things and then we're good to merge, I think.

File: test/fuzz/transformation_replace_id_with_synonym_test.cpp (hunk 13 above)
Comment (Mostafa-ashraf19) [reply test/fuzz/transformation_replace_id_with_synonym_test.cpp:2031]: Done.

File: test/fuzz/transformation_replace_id_with_synonym_test.cpp (hunk 14 above)
Comment (Mostafa-ashraf19) [reply test/fuzz/transformation_replace_id_with_synonym_test.cpp:2032]: Done.

**Review by afd — APPROVED** [review]
Looks good! Bots running.

**Comment by afd** [comment]
@Mostafa-ashraf19 The bots failed due to a death test not working as expected:

```
[ RUN      ] TransformationReplaceIdWithSynonymTest.TypesAreCompatible
../test/fuzz/transformation_replace_id_with_synonym_test.cpp:2242: Failure
Death test: TransformationReplaceIdWithSynonym::TypesAreCompatible( context.get(), SpvOpAtomicLoad, 0, int_type, uint_type)
    Result: died but not with expected error.
  Expected: contains regular expression "Invalid operand index"
```

Can you try this test out locally in a debug build and investigate, please?

**Comment by Mostafa-ashraf19** [comment]
> @Mostafa-ashraf19 The bots failed due to a death test not working as expected:
> 
> ```
> [ RUN      ] TransformationReplaceIdWithSynonymTest.TypesAreCompatible
> ../test/fuzz/transformation_replace_id_with_synonym_test.cpp:2242: Failure
> Death test: TransformationReplaceIdWithSynonym::TypesAreCompatible( context.get(), SpvOpAtomicLoad, 0, int_type, uint_type)
>     Result: died but not with expected error.
>   Expected: contains regular expression "Invalid operand index"
> ```
> 
> Can you try this test out locally in a debug build and investigate, please?

@afd I made a test locally and it's currently working well.
I think happened because the two regular expressions are different from each other (assert and ASSERT_DEATH).
Hope the bots will be well now.

**Review by afd — APPROVED** [review]
